src/LowerTriangularMatrices/lower_triangular_matrix.jl

```diff
@@ -1,117 +1,203 @@
+Base.size(L::LowerTriangularArray) = (L.m, L.n, size(L.data)[2:end]...)  # define size as matrix size (doesnt give an error, even for LowerTriangularMatrix due to tuple indexing logic)
+Base.sizeof(L::LowerTriangularArray) = sizeof(L.data)  # sizeof the underlying data vector
 
 # CREATE INSTANCES (ZEROS, ONES, UNDEF)
 for zeros_or_ones in (:zeros, :ones)
     @eval begin
-        function Base.$zeros_or_ones(::Type{LowerTriangularMatrix{T}}, m::Integer, n::Integer) where T
-            return LowerTriangularMatrix($zeros_or_ones(T, nonzeros(m, n)), m, n)
+        function Base.$zeros_or_ones(::Type{LowerTriangularArray{T,N,ArrayType}}, m::Integer, n::Integer, I::Vararg{Integer,M}) where {T,N,M,ArrayType}
+            return LowerTriangularArray(ArrayType($zeros_or_ones(T, nonzeros(m, n), I...)), m, n)
+        end
+        
+        # default CPU 
+        function Base.$zeros_or_ones(::Type{LowerTriangularArray{T}}, m::Integer, n::Integer, I::Vararg{Integer,M}) where {T,M}
+            return LowerTriangularArray($zeros_or_ones(T, nonzeros(m, n), I...), m, n)
         end
         
-        # use Float64 as default if type T not provided
+        # use Float64 as default if type T is not provided
         Base.$zeros_or_ones(::Type{LowerTriangularMatrix}, m::Integer, n::Integer) = $zeros_or_ones(LowerTriangularMatrix{Float64}, m, n)
     end
 end
 
-Base.zero(L::LowerTriangularMatrix{T}) where T = zeros(LowerTriangularMatrix{T}, size(L)...)
-Base.one(L::LowerTriangularMatrix{T}) where T = ones(LowerTriangularMatrix{T}, size(L)...)
+Base.zero(L::LowerTriangularArray{T,N,AT}) where {T,N,AT} = zeros(LowerTriangularArray{T,N,AT}, size(L)...)
+Base.one(L::LowerTriangularArray{T,N,AT}) where {T,N,AT} = zeros(LowerTriangularArray{T,N,AT}, size(L)...)
+
+function LowerTriangularArray{T,N,ArrayType}(::UndefInitializer, I::Vararg{Integer,N}) where {T,N,ArrayType<:AbstractArray{T}}
+    return LowerTriangularArray(ArrayType(undef, nonzeros(I[1], I[2]), I[3:end]...), I[1], I[2])
+end
 
+LowerTriangularArray{T,N,ArrayType}(::UndefInitializer, size::S) where {T,N,ArrayType<:AbstractArray{T}, S<:Tuple} = LowerTriangularArray{T,N,ArrayType}(undef, size...)
+   
 function LowerTriangularMatrix{T}(::UndefInitializer, m::Integer, n::Integer) where T
     return LowerTriangularMatrix(Vector{T}(undef, nonzeros(m, n)), m, n)
 end
 
-Base.randn(::Type{LowerTriangularMatrix{T}}, m::Integer, n::Integer) where T = LowerTriangularMatrix(randn(T, nonzeros(m, n)), m, n)
-Base.rand(::Type{LowerTriangularMatrix{T}}, m::Integer, n::Integer) where T = LowerTriangularMatrix(rand(T, nonzeros(m, n)), m, n)
+Base.randn(::Type{LowerTriangularArray{T}}, m::Integer, n::Integer, I::Vararg{Integer,N}) where {T,N} = LowerTriangularArray(randn(T, nonzeros(m, n), I...), m, n)
+Base.rand(::Type{LowerTriangularArray{T}}, m::Integer, n::Integer, I::Vararg{Integer,N}) where {T,N} = LowerTriangularArray(randn(T, nonzeros(m, n), I...), m, n)
```

Comment:
`rand` for `rand`?

Reply:
Replaced that with https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/503/commits/939e3584eb6a63abbf5ae1556518e78e830794c9